README.md
# CONEMO Android App

This application serves as the patient-facing Cordova portion of the NIH funded
CONEMO project.

## Build dependencies

- NodeJS
- NPM (NodeJS Package Manager)

## Installation process

[Download and install NodeJS and the Node Package Manager](http://nodejs.org/download/)

Install the application dependencies

`npm install`

`node_modules/.bin/bower install`

## Adding the Android platform

```
rm -rf platforms
mkdir www
LOCALE=es-PE API_SERVER=http://localhost:3000 \
LESSON_SERVER=http://localhost:3000 \
node_modules/.bin/cordova platform add android
```

## Adding the background mode plugin

```
LOCALE=pt-BR API_SERVER=http://localhost:3000 \
LESSON_SERVER=https://conemo.northwestern.edu \
node_modules/.bin/cordova plugin add https://github.com/katzer/cordova-plugin-background-mode.git#0.6.4
```

## Adding the start on boot plugin

```
LOCALE=pt-BR API_SERVER=http://localhost:3000 \
LESSON_SERVER=https://conemo.northwestern.edu \
node_modules/.bin/cordova plugin add https://github.com/tsubik/cordova-plugin-start-on-boot.git
```

## Running tests

`npm test`

## Building and installing the app on an Android device or emulator

```
LOCALE=es-PE API_SERVER=http://localhost:3000 \
LESSON_SERVER=http://lessons.example.com \
npm run install:android
```

## Running in the browser

The first time

```
LOCALE=es-PE API_SERVER=http://localhost:3000 \
LESSON_SERVER=http://lessons.example.com \
node_modules/.bin/cordova platform add browser
```

Then build

```
LOCALE=es-PE API_SERVER=http://localhost:3000 \
LESSON_SERVER=http://lessons.example.com \
node_modules/.bin/cordova build browser
```

And finally install (which is all you will need to run on subsequent installs)

```
LOCALE=es-PE API_SERVER=http://localhost:3000 \
LESSON_SERVER=http://lessons.example.com \
npm run install:browser
```

While running the app in the browser you may want to clear the database. To do
this run `indexedDB.deleteDatabase('conemo'); localStorage.clear()` in the
browser console.

## Building a production release

This assumes you have a build configuration file at `~/.ssh/android-build.json`

```
LOCALE=es-PE API_SERVER=http://example.com \
LESSON_SERVER=http://example.com HOCKEY_APP_ID=12345 \
node_modules/.bin/cordova build android \
--buildConfig ~/.ssh/android-build.json --release
```

## Running in an emulator

Make sure you have an SD card enabled so that the video downloads work.

If your computer is on WiFi, disable your LAN (it may be called "Thunderbolt Bridge").
Restart your virtual device and confirm that it has access to the internet.

## Testing locally with mock node server

If you'd like to run the app locally without needing to run the dashboard and
set up a participant, build the app with the mock node server,
`SERVER=http://127.0.01:1337`, as the endpoint. For example:

```
LOCALE=es-PE API_SERVER=http://localhost:3000 \
LESSON_SERVER=http://lessons.example.com \
npm run install:android
```

You will also need to run the mock server.

```
node path/to/conemo_app/test/support/dummy_server.js
```

Now, with a phone tethered to your computer and port forwarding enabled, you
can enter whatever you'd like as the configuration token.


CHANGELOG.md
# Changelog

## 1.3.31 - 2017-02-06

* update Lima content
* add boot on start plugin

## 1.3.29 - 2016-08-11

* update with latest ES training video

## 1.3.28 - 2016-08-09

* increase SP video volume
* ensure help page text fits on small screen

## 1.3.27 - 2016-08-08

* update lesson content
* fix video cropping; increase volume

## 1.3.26 - 2016-08-01

* update lesson content
* adjust markup on help page
* incorporate new Spanish training video; fix border
* update video style to embiggen

## 1.3.25 - 2016-07-11

* prevent duplicate feedback questions for activities
  (update to cache_and_sync_love library v0.2.2)
* remove blank option from pt feedback

## 1.3.24 - 2016-06-30

* update videos with higher volume
* add blank option to activity feedback

## 1.3.23 - 2016-06-30

* make checkbox labels clickable
* remove optgroup tags to fix formatting issues
* add space between activity planning yes/no btns
* add text to the help request page
* fetch and display study id on Instructions page

## 1.3.22 - 2016-06-16

* update lesson content
* add LM6 video

## 1.3.21 - 2016-06-14

* update ES videos
* update lesson content

## 1.3.20 - 2016-05-25

* pass activity planning state via element classes
* remove extraneous slide markup

## 1.3.19 - 2016-05-24

* update HockeyApp sdk; integrate metrics
* increase slide font sizes
* update content

## 1.3.17 - 2016-05-20

* ensure body class is reset on route updates

## 1.3.16 - 2016-05-20

* update content and add Portuguese video

## 1.3.15 - 2016-05-19

* update feedback slide colors
* update lesson content

## 1.3.14 - 2016-05-18

* add context-dependent slide colors

## 1.3.13 - 2016-05-18

* update translations

## 1.3.12 - 2016-05-10

* add follow up page to activity planning
* go back in the browser history on back event

## 1.3.11 - 2016-05-05

* resize text on home screen (main page)
* update lesson content
* change slide colors for activity planning/review
* use Bootstrap JS radio buttons for better UI
* resize font for activity/review selects
* save empty string for unselected activity ratings
* center activity planning/reviewing content

## 1.3.10 - 2016-04-13

* update lesson content
* use 1 as default day in study instead of 0

## 1.3.9 - 2016-04-07

* resize lesson titles for Spanish version

## 1.3.8 - 2016-04-07

* update lesson content
* resize lesson titles for Spanish version

## 1.3.7 - 2016-04-04

* fix content size for Spanish activity not done
* force Main page reload on resume

## 1.3.6 - 2016-04-04

* notify Angular of URL update rather than reloading
* apply minor style updates to activity fonts
* remove extraneous HTML from lesson slide

## 1.3.5 - 2016-03-31

* ensure videos pause on navigation
* always return to the main page on resume
* disable form inputs after saving data
* ensure videos pause on navigation
* increase Spanish version slide font size
* reduce lesson page count font size
* update lesson content
* schedule notifications based on start date

## 1.3.4 - 2016-03-25

* add blank prompt to slide form select inputs
* fix slide form input data capture
* fix data capture of activity planning and feedback
* scheduled activity feedback relative to "now"

## 1.3.3 - 2016-03-24

* update lesson content
* capture all slide form input values
* fix off-by-1 errors in slide display
* ensure unread label updates on main page

## 1.3.2 - 2016-03-23

* add lame fix for broken video functionality

## 1.3.1 - 2016-03-22

* update Portuguese lesson content
* save form on penultimate slide
* remove scrollable slide transitions
* update Spanish feedback text
* add hidden blank to activity selects

## 1.3.0 - 2016-03-21

* use Resources instead of localStorage for lessons
* use start date in Resources, not localStorage
* use global variable for 'l10n'
* remvoe unused network connection plugin and code

## 1.2.3 - 2016-03-17

* update Protuguese lesson content
* remove duplicate feedback

## 1.2.2 - 2016-03-11

* include feedback follow up ratings questions

## 1.2.1 - 2016-03-10

* style activity planning radio inputs
* make font size consistent in Portuguese app

## 1.2.0 - 2016-03-10

* update lesson content
* lock to portrait orientation
* enable customized content for activity planning
* change multi-select to checkboxes

## 1.1.9 - 2016-03-06

* translate config page
* add black background to videos
* update SP instructions
* change PT Session 2 slide 5 to be multiple choice
* remove auto numbered lesson headers
* move home button on instructions page
* add activity options
* update instructions for accuracy

## 1.1.8 - 2016-03-02

* remove training session
* change timestamp column names
* update lesson content
* fix help text PT
* add PT updated instructions
* display app version on instructions page

## 1.1.7 - 2016-02-26

* update ES content; remove ES specific CSS styles

## 1.1.6 - 2016-02-21

* update with latest lesson content
* add app version to logins data
* catch and report JS exceptions
* increase sync period to 2 minutes
* ensure lesson nav is rendered

## 1.1.5 - 2016-02-16

* fix pt-BR videos for sessions 15 and 18

## 1.1.4 - 2016-02-15

* enable crash reports to HockeyApp
* update lesson content
* update Portuguese videos
* update crosswalk and local-notification versions

## 1.1.3 - 2016-02-10

* update Cordova to v6.0.0
* install Crosswalk v16 to support Android 4.x

## 1.1.2 - 2016-02-08

* ensure session select inputs have blank option
* remove "do you want help..." q&a
* capture responses to session form inputs

## 1.1.1 - 2016-02-06

* update Lesson content
* prevent unnecessary page reload stutter
* refactoring and code cleanup
* complete activity plan/review feature in English
* fix occasional failure of lesson/session slides
* allow for building with separate server and lesson targets

## 1.1.0 - 2016-02-02

* enable background mode for data transmission
* update help contact text
* schedule lesson notifications once
* enable activity planning and reviewing

Gruntfile.js
// Generated on 2014-04-24 using generator-angular 0.7.1
/*jslint node: true */
'use strict';

// # Globbing
// for performance reasons we're only matching one level down:
// 'test/spec/{,*/}*.js'
// use this if you want to recursively match all subfolders:
// 'test/spec/**/*.js'

module.exports = function (grunt) {

  // Load grunt tasks automatically
  require('load-grunt-tasks')(grunt);

  // Time how long tasks take. Can help when optimizing build times
  require('time-grunt')(grunt);

  // Define the configuration for all the tasks
  grunt.initConfig({

    // Project settings
    yeoman: {
      // configurable paths
      app: require('./bower.json').appPath || 'app',
      dist: 'www'
    },

    // Watches files for changes and runs tasks based on the changed files
    watch: {
      js: {
        files: ['<%= yeoman.app %>/scripts/{,*/}*.js'],
        tasks: ['newer:jshint:all'],
        options: {
          livereload: true
        }
      },
      jsTest: {
        files: ['test/spec/{,*/}*.js'],
        tasks: ['newer:jshint:test', 'karma']
      },
      compass: {
        files: ['<%= yeoman.app %>/styles/{,*/}*.{scss,sass}'],
        tasks: ['compass:server', 'autoprefixer']
      },
      gruntfile: {
        files: ['Gruntfile.js']
      },
      livereload: {
        options: {
          livereload: '<%= connect.options.livereload %>'
        },
        files: [
          '<%= yeoman.app %>/{,*/}*.html',
          '.tmp/styles/{,*/}*.css',
          '<%= yeoman.app %>/images/{,*/}*.{png,jpg,jpeg,gif,webp,svg}'
        ]
      }
    },

    // The actual grunt server settings
    connect: {
      options: {
        port: 9000,
        // Change this to '0.0.0.0' to access the server from outside.
        hostname: 'localhost',
        livereload: 35729
      },
      livereload: {
        options: {
          open: true,
          base: [
            '.tmp',
            '<%= yeoman.app %>'
          ]
        }
      },
      test: {
        options: {
          port: 9001,
          base: [
            '.tmp',
            'test',
            '<%= yeoman.app %>'
          ]
        }
      },
      dist: {
        options: {
          base: '<%= yeoman.dist %>'
        }
      }
    },

    // Make sure code styles are up to par and there are no obvious mistakes
    jshint: {
      options: {
        jshintrc: '.jshintrc',
        reporter: require('jshint-stylish')
      },
      all: [
        'Gruntfile.js',
        '<%= yeoman.app %>/scripts/{,*/}*.js'
      ],
      test: {
        options: {
          jshintrc: 'test/.jshintrc'
        },
        src: ['test/spec/{,*/}*.js']
      }
    },

    // Empties folders to start fresh
    clean: {
      dist: {
        files: [{
          dot: true,
          src: [
            '.tmp',
            '<%= yeoman.dist %>/*',
            '!<%= yeoman.dist %>/.git*'
          ]
        }]
      },
      server: '.tmp'
    },

    // Add vendor prefixed styles
    autoprefixer: {
      options: {
        browsers: ['last 1 version']
      },
      dist: {
        files: [{
          expand: true,
          cwd: '.tmp/styles/',
          src: '{,*/}*.css',
          dest: '.tmp/styles/'
        }]
      }
    },

    // Automatically inject Bower components into the app
    'bower-install': {
      app: {
        html: '<%= yeoman.app %>/index.html',
        ignorePath: '<%= yeoman.app %>/'
      }
    },




    // Compiles Sass to CSS and generates necessary files if requested
    compass: {
      options: {
        sassDir: '<%= yeoman.app %>/styles',
        cssDir: '.tmp/styles',
        generatedImagesDir: '.tmp/images/generated',
        imagesDir: '<%= yeoman.app %>/images',
        javascriptsDir: '<%= yeoman.app %>/scripts',
        fontsDir: '<%= yeoman.app %>/styles/fonts',
        importPath: '<%= yeoman.app %>/../bower_components',
        httpImagesPath: '/images',
        httpGeneratedImagesPath: '/images/generated',
        httpFontsPath: '/styles/fonts',
        relativeAssets: false,
        assetCacheBuster: false,
        raw: 'Sass::Script::Number.precision = 10\n'
      },
      dist: {
        options: {
          generatedImagesDir: '<%= yeoman.dist %>/images/generated'
        }
      },
      server: {
        options: {
          debugInfo: true
        }
      }
    },

    // Renames files for browser caching purposes
    rev: {
      dist: {
        files: {
          src: [
            '<%= yeoman.dist %>/scripts/{,*/}*.js',
            '<%= yeoman.dist %>/styles/{,*/}*.css',
            '<%= yeoman.dist %>/images/{,*/}*.{png,jpg,jpeg,gif,webp,svg}',
            '<%= yeoman.dist %>/styles/fonts/*'
          ]
        }
      }
    },

    // Reads HTML for usemin blocks to enable smart builds that automatically
    // concat, minify and revision files. Creates configurations in memory so
    // additional tasks can operate on them
    useminPrepare: {
      html: '<%= yeoman.app %>/index.html',
      options: {
        dest: '<%= yeoman.dist %>',
        flow: {
          html: {
            steps: {
              js: ['concat'],
              css: ['cssmin']
            },
            post: {}
          }
        }
      }
    },

    // Performs rewrites based on rev and the useminPrepare configuration
    usemin: {
      html: ['<%= yeoman.dist %>/{,*/}*.html'],
      css: ['<%= yeoman.dist %>/styles/{,*/}*.css'],
      options: {
        assetsDirs: ['<%= yeoman.dist %>']
      }
    },

    // The following *-min tasks produce minified files in the dist folder
    imagemin: {
      dist: {
        files: [{
          expand: true,
          cwd: '<%= yeoman.app %>/images',
          src: '{,*/}*.{png,jpg,jpeg,gif}',
          dest: '<%= yeoman.dist %>/images'
        }]
      }
    },
    svgmin: {
      dist: {
        files: [{
          expand: true,
          cwd: '<%= yeoman.app %>/images',
          src: '{,*/}*.svg',
          dest: '<%= yeoman.dist %>/images'
        }]
      }
    },
    htmlmin: {
      dist: {
        options: {
          collapseWhitespace: true,
          collapseBooleanAttributes: true,
          removeCommentsFromCDATA: true,
          removeOptionalTags: true
        },
        files: [{
          expand: true,
          cwd: '<%= yeoman.dist %>',
          src: ['*.html', 'views/{,*/}*.html'],
          dest: '<%= yeoman.dist %>'
        }]
      }
    },

    // Allow the use of non-minsafe AngularJS files. Automatically makes it
    // minsafe compatible so Uglify does not destroy the ng references
    ngmin: {
      dist: {
        files: [{
          expand: true,
          cwd: '.tmp/concat/scripts',
          src: '*.js',
          dest: '.tmp/concat/scripts'
        }]
      }
    },

    // Replace Google CDN references
    cdnify: {
      dist: {
        html: ['<%= yeoman.dist %>/*.html']
      }
    },

    // Copies remaining files to places other tasks can use
    copy: {
      dist: {
        files: [{
          expand: true,
          dot: true,
          cwd: '<%= yeoman.app %>',
          dest: '<%= yeoman.dist %>',
          src: [
            '*.{ico,png,txt}',
            '.htaccess',
            '*.html',
            'views/{,*/}*.html',
            '/bower_components/**/*',
            'images/{,*/}*.{webp}',
            'fonts/*'
          ]
        }, {
          expand: true,
          cwd: '.tmp/images',
          dest: '<%= yeoman.dist %>/images',
          src: ['generated/*']
        }]
      },
      styles: {
        expand: true,
        cwd: '<%= yeoman.app %>/styles',
        dest: '.tmp/styles/',
        src: '{,*/}*.css'
      }
    },

    // Run some tasks in parallel to speed up the build process
    concurrent: {
      server: [
        'compass:server'
      ],
      test: [
        'compass'
      ],
      dist: [
        'compass:dist',
        'imagemin',
        'svgmin'
      ]
    },

    // By default, your `index.html`'s <!-- Usemin block --> will take care of
    // minification. These next options are pre-configured if you do not wish
    // to use the Usemin blocks.
    // cssmin: {
    //   dist: {
    //     files: {
    //       '<%= yeoman.dist %>/styles/main.css': [
    //         '.tmp/styles/{,*/}*.css',
    //         '<%= yeoman.app %>/styles/{,*/}*.css'
    //       ]
    //     }
    //   }
    // },
    // uglify: {
    //   dist: {
    //     files: {
    //       '<%= yeoman.dist %>/scripts/scripts.js': [
    //         '<%= yeoman.dist %>/scripts/scripts.js'
    //       ]
    //     }
    //   }
    // },
    // concat: {
    //   dist: {}
    // },

    // Test settings
    karma: {
      unit: {
        configFile: 'karma.conf.js',
        singleRun: true
      }
    }
  });


  grunt.registerTask('serve', function (target) {
    if (target === 'dist') {
      return grunt.task.run(['build', 'connect:dist:keepalive']);
    }

    grunt.task.run([
      'clean:server',
      'bower-install',
      'concurrent:server',
      'autoprefixer',
      'connect:livereload',
      'watch'
    ]);
  });

  grunt.registerTask('server', function () {
    grunt.log.warn('The `server` task has been deprecated. Use `grunt serve` to start a server.');
    grunt.task.run(['serve']);
  });

  grunt.registerTask('test', [
    'clean:server',
    'concurrent:test',
    'autoprefixer',
    'connect:test',
    'karma'
  ]);

  grunt.registerTask('build', [
    'clean:dist',
    'useminPrepare',
    'concurrent:dist',
    'autoprefixer',
    'concat',
    'compass',
    'ngmin',
    'copy:dist',
    'cdnify',
    'cssmin',
    //'uglify',
    'rev',
    'usemin',
    //'htmlmin'
  ]);

  grunt.registerTask('default', [
    'newer:jshint',
    'test',
    'build'
  ]);
};
